src/highdicom/io.py

```diff
@@ -0,0 +1,780 @@
+_JPEG2000_SOC_MARKER = b'\xFF\x4F'
+_JPEG2000_EOC_MARKER = b'\xFF\xD9'
+_START_MARKERS = {_JPEG_SOI_MARKER, _JPEG2000_SOC_MARKER}
+_END_MARKERS = {_JPEG_EOI_MARKER, _JPEG2000_EOC_MARKER}
+
+
+def _get_bot(fp: DicomFile, number_of_frames: int) -> List[int]:
+    """Tries to read the value of the Basic Offset Table (BOT) item and builds
+    it case it is not found.
+
+    Parameters
+    ----------
+    fp: pydicom.filebase.DicomFile
+        Pointer for DICOM PS3.10 file stream positioned at the first byte of
+        the Pixel Data element
+    number_of_frames: int
+        Number of frames contained in the Pixel Data element
+
+    Returns
+    -------
+    List[int]
+        Offset of each Frame item in bytes from the first byte of the Pixel Data
+        element following the BOT item
+
+    Note
+    ----
+    Moves the pointer to the first byte of the open file following the BOT item
+    (the first byte of the first Frame item).
+
+    """
+    logger.debug('read Basic Offset Table')
+    basic_offset_table = _read_bot(fp)
+
+    first_frame_offset = fp.tell()
+    tag = TupleTag(fp.read_tag())
+    if int(tag) != ItemTag:
+        raise ValueError('Reading of Basic Offset Table failed')
+    fp.seek(first_frame_offset, 0)
+
+    # Basic Offset Table item must be present, but it may be empty
+    if len(basic_offset_table) == 0:
+        logger.debug('Basic Offset Table item is empty')
+    if len(basic_offset_table) != number_of_frames:
+        logger.debug('build Basic Offset Table item')
+        basic_offset_table = _build_bot(
+            fp,
+            number_of_frames=number_of_frames
+        )
+
+    return basic_offset_table
+
+
+def _read_bot(fp: DicomFile) -> List[int]:
+    """Reads the Basic Offset Table (BOT) item of an encapsulated Pixel Data
+    element.
+
+    Parameters
+    ----------
+    fp: pydicom.filebase.DicomFile
+        Pointer for DICOM PS3.10 file stream positioned at the first byte of
+        the Pixel Data element
+
+    Returns
+    -------
+    List[int]
+        Offset of each Frame item in bytes from the first byte of the Pixel Data
+        element following the BOT item
+
+    Note
+    ----
+    Moves the pointer to the first byte of the open file following the BOT item
+    (the first byte of the first Frame item).
+
+    Raises
+    ------
+    IOError
+        When file pointer is not positioned at first byte of Pixel Data element
+
+    """
+    tag = TupleTag(fp.read_tag())
+    if int(tag) not in _PIXEL_DATA_TAGS:
+        raise IOError(
+            'Expected file pointer at first byte of Pixel Data element.'
+        )
+    # Skip Pixel Data element header (tag, VR, length)
+    pixel_data_element_value_offset = data_element_offset_to_value(
+        fp.is_implicit_VR, 'OB'
+    )
+    fp.seek(pixel_data_element_value_offset - 4, 1)
+    is_empty, offsets = get_frame_offsets(fp)
+    return offsets
+
+
+def _build_bot(fp: DicomFile, number_of_frames: int) -> List[int]:
```

Comment:
Consider contributing to pydicom.encaps instead of here?

Reply:
@scaramallion would you be open to including this in `pydicom.encaps`, for example as `pydicom.encaps.compute_frame_offsets()` (to match `pydicom.encaps.get_frame_offsets()`)?